lib/posix/pthread_mutex.c

```diff
@@ -9,16 +9,199 @@
 #include <wait_q.h>
 #include <posix/pthread.h>
 
+#define MUTEX_MAX_REC_LOCK 32767
+/* Memory pool for pthread space */
+K_MEM_POOL_DEFINE(posix_mutex_sem, sizeof(struct k_sem),
+		  sizeof(struct k_sem), 32, 4);
+
+/*
+ *  Default mutex attrs.
+ */
+static const pthread_mutexattr_t def_attr = {
+	.type = PTHREAD_MUTEX_DEFAULT,
+};
+
+static int acquire_mutex(pthread_mutex_t *m, int timeout)
+{
+	int rc = 0, type, count;
+	pthread_t owner;
+	int key = irq_lock();
+
+	owner = m->owner;
+	type = m->type;
+	count = m->lock_count;
+	irq_unlock(key);
+
+	if (owner == pthread_self()) {
+		if (type == PTHREAD_MUTEX_ERRORCHECK) {
+			return EDEADLK;
+		}
+	}
+
+	if (type == PTHREAD_MUTEX_RECURSIVE) {
+
+		if ((owner == pthread_self()) || (count == 0)) {
+			if (count >= MUTEX_MAX_REC_LOCK) {
+				return EAGAIN;
+			}
+			rc = -k_sem_take(m->sem, timeout);
+			count++;
+		} else {
+			rc = EPERM;
+		}
+	} else {
+		rc = k_sem_take(m->sem, timeout);
+		count = 1;
+	}
+
+	if (rc == 0) {
+		key = irq_lock();
+		m->owner = pthread_self();
+		m->lock_count = count;
+		irq_unlock(key);
+	}
+
+	return rc;
+}
+
+/**
+ * @brief Lock POSIX mutex with non-blocking call.
+ *
+ * See IEEE 1003.1
+ */
 int pthread_mutex_trylock(pthread_mutex_t *m)
 {
-	int key = irq_lock(), ret = EBUSY;
+	return acquire_mutex(m, K_NO_WAIT);
+}
+
+/**
+ * @brief Intialize POSIX mutex.
+ *
+ * See IEEE 1003.1
+ */
+int pthread_mutex_init(pthread_mutex_t *m,
+				     const pthread_mutexattr_t *attr)
+{
+	const pthread_mutexattr_t *mattr;
+
+	if (k_mem_pool_alloc(&posix_mutex_sem, &m->block,
+			sizeof(struct k_sem), K_FOREVER) == 0) {
+		memset(m->block.data, 0, sizeof(struct k_sem));
+	}
+
+	m->sem = m->block.data;
+
+	m->owner = NULL;
+	m->lock_count = 0;
 
-	if (m->sem->count) {
-		m->sem->count = 0;
-		ret = 0;
+	mattr = (attr == NULL) ? &def_attr : attr;
+
+	m->type = mattr->type;
+
+	if ((m->type == PTHREAD_MUTEX_NORMAL) ||
+		(m->type == PTHREAD_MUTEX_ERRORCHECK)) {
+		k_sem_init(m->sem, 1, 1);
+	} else {
+		k_sem_init(m->sem, MUTEX_MAX_REC_LOCK, MUTEX_MAX_REC_LOCK);
 	}
 
+	return 0;
+}
+
+/**
+ * @brief Lock POSIX mutex with blocking call.
+ *
+ * See IEEE 1003.1
+ */
+int pthread_mutex_lock(pthread_mutex_t *m)
+{
+	return acquire_mutex(m, K_FOREVER);
+}
+
+/**
+ * @brief Unlock POSIX mutex.
+ *
+ * See IEEE 1003.1
+ */
+int pthread_mutex_unlock(pthread_mutex_t *m)
+{
+	pthread_t owner;
+	int count, key = irq_lock();
+
+	owner = m->owner;
+	count = m->lock_count;
+	irq_unlock(key);
+
+	if (owner != pthread_self()) {
+		return EPERM;
+	}
+
+	if (count == 0) {
+		return EINVAL;
+	}
+
```

Comment:
There's a race here too.  Consider two threads calling pthread_mutex_unlock() incorrectly, on a lock that was locked only once.  Both of them get to this location where I added the comment, decide to decrement the lock, and continue on.  Now both threads have completed successfully when one of them should have returned an error.   And worse, lock_count is now -1 and future calls are probably going to explode in hilarious ways.